_WKHitTestResult should expose an elementType property

jeffmapple opened the pull request:
#### 824b10f0e852a126afc79ec2093271a5eca869b6
<pre>
_WKHitTestResult should expose an elementType property
<a href="https://bugs.webkit.org/show_bug.cgi?id=262137">https://bugs.webkit.org/show_bug.cgi?id=262137</a>
rdar://116078370

Reviewed by Aditya Keerthi and Wenson Hsieh.

Add a elementType property to _WKHitTestResult and plumbing to WebCore::HitTestResult and
WebKit::WebHitTestResultData to support it.

* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::hasMediaElement const):
* Source/WebCore/rendering/HitTestResult.h:
Added hasMediaElement().
* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.h:
Added _WKHitTestResultElementType and elementType property.
* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.mm:
(-[_WKHitTestResult elementType]):
* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::elementTypeFromHitTestResult):
Added, derives WebHitTestResultData::ElementType from WebCore::HitTestResult.
(WebKit::WebHitTestResultData::WebHitTestResultData):
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
Add allowed values for WebHitTestResultData::ElementType and add elementType to WebHitTestResultData.
* Source/WebKit/UIProcess/API/APIHitTestResult.h:
(API::HitTestResult::elementType const):
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST):
Added HitTestResultElementTypeNone and HitTestResultElementTypeVideo tests to verify elementType
property is set correctly on _WKHitTestResult.

Canonical link: <a href="https://commits.webkit.org/268545@main">https://commits.webkit.org/268545@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab11aab8d0766531a1c3f65a7c23c0715b1c643e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20003 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20428 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21051 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21898 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/18680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20234 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23678 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20579 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20223 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20165 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17390 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22750 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17335 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18186 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/24449 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18412 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18362 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22438 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18953 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18136 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4786 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22486 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18779 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jeffmapple commented on Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.h:
```diff
@@ -45,6 +51,8 @@ WK_CLASS_AVAILABLE(macos(10.12), ios(16.0))
 
 @property (nonatomic, readonly) CGRect elementBoundingBox;
 
+@property (nonatomic, readonly) _WKHitTestResultMediaType mediaType;
```
Just realized I forgot to add `WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA))`, will fix.

pxlcoder commented on Source/WebKit/Shared/WebHitTestResultData.h:
```diff
@@ -84,6 +84,8 @@ struct WebHitTestResultData {
     bool isOverTextInsideFormControlElement;
     bool isDownloadableMedia;
+    bool hasMediaElement;
+    bool mediaIsVideo;
```
Can we make the implementation match the SPI surface – using an enumeration rather than two booleans?

We may want to extend this in the future. Something that comes to mind is a way to expose that a `<model>` element is being context clicked.

jeffmapple replied:
Should the `_WKHitTestResultMediaType` and `mediaType` names be more generic in `_WKHitTestResult`? Maybe `_WKHitTestResultElementType` and `elementType`?

pxlcoder replied:
That sounds good to me!

jeffmapple committed:
_WKHitTestResult should expose an elementType property
https://bugs.webkit.org/show_bug.cgi?id=262137
rdar://116078370

Reviewed by Aditya Keerthi and Wenson Hsieh.

Add a elementType property to _WKHitTestResult and plumbing to WebCore::HitTestResult and
WebKit::WebHitTestResultData to support it.

* Source/WebCore/rendering/HitTestResult.cpp:
(WebCore::HitTestResult::hasMediaElement const):
* Source/WebCore/rendering/HitTestResult.h:
Added hasMediaElement().
* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.h:
Added _WKHitTestResultElementType and elementType property.
* Source/WebKit/Shared/API/Cocoa/_WKHitTestResult.mm:
(-[_WKHitTestResult elementType]):
* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::elementTypeFromHitTestResult):
Added, derives WebHitTestResultData::ElementType from WebCore::HitTestResult.
(WebKit::WebHitTestResultData::WebHitTestResultData):
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
Add allowed values for WebHitTestResultData::ElementType and add elementType to WebHitTestResultData.
* Source/WebKit/UIProcess/API/APIHitTestResult.h:
(API::HitTestResult::elementType const):
* Tools/TestWebKitAPI/Tests/mac/ContextMenuTests.mm:
(TestWebKitAPI::TEST):
Added HitTestResultElementTypeNone and HitTestResultElementTypeVideo tests to verify elementType
property is set correctly on _WKHitTestResult.

Canonical link: https://commits.webkit.org/268545@main
